File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/levels/objects/Barrel.java
package com.nyrds.pixeldungeon.levels.objects;

import androidx.annotation.Keep;

import com.nyrds.pixeldungeon.ml.R;
import com.nyrds.platform.audio.Sample;
import com.nyrds.platform.util.StringsManager;
import com.nyrds.util.ModdingMode;
import com.watabou.pixeldungeon.Assets;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.blobs.Blob;
import com.watabou.pixeldungeon.actors.blobs.LiquidFlame;
import com.watabou.pixeldungeon.levels.Level;
import com.watabou.pixeldungeon.scenes.GameScene;

import org.json.JSONObject;

/**
 * Created by mike on 01.07.2016.
 */
public class Barrel extends LevelObject {

	private boolean burned = false;

	@Keep
	public Barrel() {
		this(-1);
	}

	public Barrel(int pos) {
		super(pos);
		textureFile = "levelObjects/barrels.png";
	}

	@Override
	void setupFromJson(Level level, JSONObject obj) {
	}

	@Override
	public boolean pushable(Char hero) {
		return true;
	}

	@Override
	public void burn() {

		if (burned) {
			return;
		}

		burned = true;

		sprite.playAnim(10, false, () -> remove(), image() + 0, image() + 1, image() + 2, image() + 3, image() + 4);


		Sample.INSTANCE.play(Assets.SND_EXPLOSION);

		LiquidFlame fire = Blob.seed(getPos(), 10, LiquidFlame.class);
		GameScene.add(fire);
	}

	@Override
	public void bump(Presser presser) {
		burn();
	}

	@Override
	public String desc() {

		if (ModdingMode.isHalloweenEvent()) {
            return StringsManager.getVar(R.string.Barrel_Pumpkin_Desc);
        } else {
            return StringsManager.getVar(R.string.Barrel_Desc);
        }

	}

	@Override
	public String name() {
		if (ModdingMode.isHalloweenEvent()) {
            return StringsManager.getVar(R.string.Barrel_Pumpkin_Name);
        } else {
            return StringsManager.getVar(R.string.Barrel_Name);
        }
	}

	@Override
	public int image() {
		if (ModdingMode.isHalloweenEvent()) {
			return 0;
		} else {
			return 8;
		}
	}

	@Override
	public boolean affectLevelObjects() {
		return true;
	}

	@Override
	public boolean nonPassable(Char ch) {
		if(ch.valid()) {
			return true;
		}
		return false;
	}
}

File: scripts/objects/barrel.lua
---
--- Barrel, data-served (batch 21). Burns once into a LiquidFlame seed; the
--- burn-down animation comes from the def ("burn"), removal rides its
--- completion hook.
---

local RPD = require "scripts/lib/commonClasses"

local object = require "scripts/lib/object"


return object.init{

    stepOn = function(self, object, hero)
        return true
    end,

    pushable = function(self, object, hero)
        return true
    end,

    nonPassable = function(self, object, ch)
        return ch:valid()
    end,

    affectItems = function(self, object)
        return true
    end,

    affectLevelObjects = function(self, object)
        return true
    end,

    flammable = function(self, object)
        return true
    end,

    burn = function(self, object)
        local st = self:restoreData()
        if st.burned then
            return
        end
        st.burned = true
        self:storeData(st)

        object:playObjectAnim("burn", "onBurned")
        RPD.playSound("snd_explosion")
        RPD.placeBlob(RPD.Blobs.LiquidFlame, object:getPos(), 10)
    end,

    onBurned = function(self, object)
        object:remove()
    end,

    image = function(self, object, level)
        if RPD.ModdingBase:isHalloweenEvent() then
            return 0
        end
        return 8
    end,

    name = function(self, object, level)
        if RPD.ModdingBase:isHalloweenEvent() then
            return RPD.textById("Barrel_Pumpkin_Name")
        end
        return RPD.textById("Barrel_Name")
    end,

    info = function(self, object, level)
        if RPD.ModdingBase:isHalloweenEvent() then
            return RPD.textById("Barrel_Pumpkin_Desc")
        end
        return RPD.textById("Barrel_Desc")
    end
}
